src/Compilers/CSharp/Portable/Syntax/CSharpNonRecursiveSyntaxRewriter.cs

```diff
@@ -0,0 +1,253 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Represents a non-recursive visitor which descends an entire <see cref="CSharpSyntaxNode"/> graph and
+    /// may replace or remove visited SyntaxNodes in depth-first order.
+    /// </summary>
+    public class CSharpNonRecursiveSyntaxRewriter : CSharpSyntaxVisitor<SyntaxNode>
+    {
+        private readonly Stack<SyntaxNodeOrToken> _undeconstructedStack;
+        private readonly Stack<UntransformedNode> _untransformedStack;
+        private readonly Stack<SyntaxNodeOrToken> _transformedStack;
+        private readonly NodeDeconstructor _deconstructor;
+        private readonly NodeReassembler _reassembler;
+        private readonly TriviaRewriter _trivializer;
+        private readonly List<SyntaxNodeOrToken> _children;
+
+        protected CSharpNonRecursiveSyntaxRewriter(bool visitIntoStructuredTrivia = false)
+        {
+            _undeconstructedStack = new Stack<SyntaxNodeOrToken>();
+            _transformedStack = new Stack<SyntaxNodeOrToken>();
+            _untransformedStack = new Stack<UntransformedNode>();
+            _deconstructor = new NodeDeconstructor();
+            _reassembler = new NodeReassembler();
+            _trivializer = new TriviaRewriter(this, visitIntoStructuredTrivia);
+            _children = new List<SyntaxNodeOrToken>();
+        }
+
+        protected SyntaxNode Original { get; private set; }
+
+        protected virtual bool Skip(SyntaxNodeOrToken nodeOrToken, out SyntaxNodeOrToken rewriten)
```

Comment:
>Skip( [](start = 31, length = 5)

Why does skipping a node include rewriting it?  What is being skipped?  If its skipping the rewriting of its children, possibly it could be named better.  ShouldRewriteChildren?

I think there should be equivalent checks for token and trivia types too.  One of the common things to do with the existing recursive rewriter is to limit the rewrites of children when you know they are outside of the span range that you want to affect.